keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -409,16 +409,42 @@ class SqlArtifactRepository(
   ) {
     val environment = deliveryConfig.environmentNamed(targetEnvironment)
     val environmentUid = deliveryConfig.getUidFor(environment)
+    val now = clock.instant()
     sqlRetry.withRetry(WRITE) {
-      jooq
-        .insertInto(ENVIRONMENT_ARTIFACT_VERSIONS)
-        .set(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID, environmentUid)
-        .set(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID, artifact.uid)
-        .set(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION, version)
-        .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, DEPLOYING.name)
-        .onDuplicateKeyUpdate()
-        .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, DEPLOYING.name)
-        .execute()
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+        txn.select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION)
+          .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+          .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
+          .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifact.uid))
+          .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(DEPLOYING.name))
+          .forUpdate()
+          .fetch()
+          .also {
+            it.forEach { (stuckVersion) ->
+              log.error("Stuck deploying version $stuckVersion for artifact '${artifact.reference}' in delivery config ${deliveryConfig.name} found when deploying version $version")
+              txn.update(ENVIRONMENT_ARTIFACT_VERSIONS)
+                .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, SKIPPED.name)
+                .set(ENVIRONMENT_ARTIFACT_VERSIONS.REPLACED_BY, version)
+                .set(ENVIRONMENT_ARTIFACT_VERSIONS.REPLACED_AT, now.toLocal())
+                .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifact.uid))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.eq(DEPLOYING.name))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(stuckVersion))
+                .execute()
+            }
+          }
+
+        txn
+          .insertInto(ENVIRONMENT_ARTIFACT_VERSIONS)
+          .set(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID, environmentUid)
+          .set(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID, artifact.uid)
+          .set(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION, version)
+          .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, DEPLOYING.name)
+          .onDuplicateKeyUpdate()
+          .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, DEPLOYING.name)
```

Comment:
I don't quite understand how `onDuplicateKeyUpdate()` works. How do you know which columns to update here?

Reply:
The `.onDuplicateKeyUpdate()` is part of the `.insertInto` statement. So, if a row already exists with a key that matches what that insert would generate, instead we update that record and set the specific fields we list